src/Sarif/Baseline/V2/WhatComparer.cs

```diff
@@ -0,0 +1,192 @@
+        public static bool MatchesWhat(ExtractedResult left, ExtractedResult right)
+        {
+            if (left?.Result == null || right?.Result == null) { return false; }
+
+            // Match Guid
+            if (left.Result.Guid != null)
+            {
+                if (string.Equals(left.Result.Guid, right.Result.Guid))
+                {
+                    return true;
+                }
+
+                // Non-Match doesn't force false, as GUIDs are often re-generated by tools.
+            }
+
+            // Match Fingerprints (any one match is a match)
+            if (left.Result.Fingerprints != null && right.Result.Fingerprints != null)
+            {
+                int correspondingFingerprintCount = 0;
+
+                foreach (var fingerprint in left.Result.Fingerprints)
+                {
+                    if (right.Result.Fingerprints.TryGetValue(fingerprint.Key, out string otherFingerprint))
+                    {
+                        correspondingFingerprintCount++;
+
+                        if (string.Equals(fingerprint.Value, otherFingerprint))
+                        {
+                            return true;
+                        }
+                    }
+                }
+
+                // Force non-match if there were fingerprints but none of them matched
+                if (correspondingFingerprintCount > 0) { return false; }
+            }
+
+            // Match PartialFingerprints (50% must match)
+            if (left.Result.PartialFingerprints != null && right.Result.PartialFingerprints != null)
+            {
+                int correspondingFingerprintCount = 0;
+                int matchCount = 0;
+
+                foreach (var fingerprint in left.Result.PartialFingerprints)
+                {
+                    if (right.Result.PartialFingerprints.TryGetValue(fingerprint.Key, out string otherFingerprint))
+                    {
+                        correspondingFingerprintCount++;
+
+                        if (string.Equals(fingerprint.Value, otherFingerprint))
+                        {
+                            matchCount++;
+                        }
+                    }
+                }
+
+                if (correspondingFingerprintCount > 0)
+                {
+                    // If fingerprints were found, return true when at least half of them matched, false otherwise
+                    return (matchCount > 0 && matchCount * 2 >= left.Result.PartialFingerprints.Count);
+                }
+            }
+
+            // Match Properties (50% must match)
+            if (left.Result.Properties != null && right.Result.Properties != null)
+            {
+                int correspondingPropertyCount = 0;
+                int matchCount = 0;
+
+                foreach (var property in left.Result.Properties)
+                {
+                    if (left.Result.TryGetProperty(property.Key, out string leftValue)
```

Comment:
>out string leftValue [](start = 65, length = 20)

This looks wrong -- property bag elements don't have to be strings; they can be any object. Let me look into this...

Reply:
Yes, this is a bug. If the property bag has an entry with the specified key, but the corresponding value is not a string, the SDK will `throw new InvalidOperationException(SdkResources.CallGenericGetProperty)`.

I have to think about what to do here.

---
In reply to: [296411642](https://github.com/microsoft/sarif-sdk/pull/1539#discussion_r296411642) [](ancestors = 296411642)